jwst/flatfield/flat_field_step.py

```diff
@@ -61,111 +74,48 @@ def process(self, input):
-            if self.d_flat_filename == 'N/A':
-                d_flat_model = None
+        # Get reference file paths
+        reference_file_names = {}
+        for reftype in self.reference_file_types:
+            reffile = self.get_reference_file(input_model, reftype)
+            reference_file_names[reftype] = reffile if reffile != 'N/A' else None
+
+        # Define mapping between reftype and datamodel type
+        model_type = dict(
+            flat=datamodels.FlatModel,
+            fflat=datamodels.NirspecFlatModel,
+            sflat=datamodels.NirspecFlatModel,
+            dflat=datamodels.NirspecFlatModel,
+            )
+        if exposure_type == "NRS_MSASPEC":
+            model_type["fflat"] = datamodels.NirspecQuadFlatModel
+
+        # Open the relevant reference files as datamodels
+        reference_file_models = {}
+        for reftype, reffile in reference_file_names.items():
+            if reffile is not None:
+                reference_file_models[reftype] = model_type[reftype](reffile)
+                self.log.debug('Using %s reference file: %s', reftype.upper(), reffile)
             else:
-                d_flat_model = datamodels.NirspecFlatModel(self.d_flat_filename)
-        else:
-            self.log.debug('Opening flat as FlatModel')
-            flat_model = datamodels.FlatModel(self.flat_filename)
-            f_flat_model = None
-            s_flat_model = None
-            d_flat_model = None
-            self.flat_suffix = None
+                reference_file_models[reftype] = None
 
         # Do the flat-field correction
-        (output_model, interpolated_flats) = \
-                flat_field.do_correction(input_model, flat_model,
-                                         f_flat_model, s_flat_model,
-                                         d_flat_model, self.flat_suffix)
+        output_model, interpolated_flats = flat_field.do_correction(
+            input_model,
+            **reference_file_models,
+            )
 
-        # Close the inputs
+        # Close the input and reference files
         input_model.close()
-        if is_NRS_spectrographic:
-            if f_flat_model is not None:
-                f_flat_model.close()
-            if s_flat_model is not None:
-                s_flat_model.close()
-            if d_flat_model is not None:
-                d_flat_model.close()
-        else:
-            flat_model.close()
-
-        if interpolated_flats is not None:
-            self.log.info("Writing interpolated flat fields.")
-            self.save_model(interpolated_flats, self.flat_suffix)
+        try:
+            for model in reference_file_models.values():
+                model.close()
+        except AttributeError:
+            pass
+
+        if self.save_interpolated_flat and interpolated_flats is not None:
+            self.log.info("Writing interpolated flat field.")
+            self.save_model(interpolated_flats, suffix=self.flat_suffix)
```

Comment:
Should this be just `flat_suffix` rather than `self.flat_suffix`?

Reply:
Since this in the `self.process()` method, that method can only refer to class attributes via `self`.  I.e. they do not have a global namespace.  I suppose I could be a bit more explicit and actually set `self.flat_suffix` in a new `__init__()` method for the `FlatFieldStep` subclass, but given this is meant to be a "constant" of the class, I felt that wasn't necessary.  Thoughts?
```
In [11]: from jwst.flatfield import FlatFieldStep

In [12]: FlatFieldStep.flat_suffix
Out[12]: 'interpolatedflat'

In [13]: step = FlatFieldStep()

2019-05-21 13:23:09,123 - stpipe.FlatFieldStep - INFO - FlatFieldStep instance created.
In [14]: step.flat_suffix
Out[14]: 'interpolatedflat'
```